src/Sylius/Bundle/CoreBundle/Controller/ProductTaxonController.php

```diff
@@ -37,23 +37,25 @@ public function updatePositionsAction(Request $request): Response
         }
 
         if (in_array($request->getMethod(), ['POST', 'PUT', 'PATCH'], true) && null !== $productTaxons) {
+            /** @var Session $session */
+            $session = $request->getSession();
+
             /** @var ProductTaxonInterface $productTaxon */
-            foreach ($productTaxons as $productTaxon) {
-                if (!is_numeric($productTaxon['position'])) {
-                    throw new HttpException(
-                        Response::HTTP_BAD_REQUEST,
-                        sprintf('The productTaxon position "%s" is invalid.', $productTaxon['position'])
-                    );
+            foreach ($productTaxons as $id => $position) {
+                if (!is_numeric($position)) {
+                    $session->getFlashBag()->add('error', sprintf('The position "%s" is invalid.', $position));
+
+                    return $this->redirectHandler->redirectToReferer($configuration);
                 }
 
                 /** @var ProductTaxonInterface $productTaxonFromBase */
-                $productTaxonFromBase = $this->repository->findOneBy(['id' => $productTaxon['id']]);
-                $productTaxonFromBase->setPosition((int) $productTaxon['position']);
-
-                $this->manager->flush();
+                $productTaxonFromBase = $this->repository->findOneBy(['id' => $id]);
+                $productTaxonFromBase->setPosition((int) $position);
             }
+
+            $this->manager->flush();
         }
 
-        return new JsonResponse();
```

Comment:
On the one hand, we kept the contract, but if someone was using this action it may stop working as expected. Same as with routes, we should probably deprecate this class and create a new implementation